experimental/csf_proficiency/elo_modeling.rb

```diff
@@ -0,0 +1,143 @@
+# Fixing a concept C and restricting to levels testing (only) C, the ELO Model
+# we use assumes the probability of success is a function of the difficulty D of
+# the level and the actual proficiency AP of the user. This probability is given
+# by the function
+#
+#   Prob_C(D, AP) = 1 / (1 + EXP(alpha * (AP - D))
+#
+# where alpha is some pre-determined constant.
+#
+#
+# Given progress (via a UserLevel) on a level with difficulty D for concept C,
+# we update the measured proficiency MP by the function
+#
+#   MP_new = MP_old + kappa * (actual - Prob_C(D, MP_old))
+#
+# where kappa is some pre-determined constant and actual is the outcome in the
+# UserLevel.
+
+require 'csv'
+
+# Picked so that Prob_C(D, AP) = 0.9 anytime AP - D = 1.
+ALPHA = -Math.log(9)
+# Picked magically so that both old (via MP_old) and new (via actual) progress
+# have significant impact on measured proficiency (via MP_new).
+KAPPA = 0.5
+# Mapping from concepts to array index position.
+CONCEPT_TO_INDEX_HASH = {
+  sequencing: 0,
+  debugging: 1,
+  repeat_loops: 2,
+  repeat_until_while: 3,
+  for_loops: 4,
+  events: 5,
+  variables: 6,
+  functions: 7,
+  functions_with_params: 8,
+  conditionals: 9
+}.freeze
+
+# @param file [String] filepath to a CSV containing UserLevel summary data.
+#   Each for of the CSV should contain a user_id, level_id, best_result,
+#   num attempts, num_hints, and num_authored_hints.
+# @return [Hash[Integer, Hash]] hash mapping user_ids to hashes, themselves
+#   mapping level_id to user_level information.
+def read_data_file(file)
+  data = Hash.new {|h, k| h[k] = Hash.new {|hp, kp| hp[kp] = {}}}
+  CSV.foreach(file) do |row|
+    row = row.map(&:to_i)
+    data[row[0]][row[1]] = row[2..5]
+  end
+  data
+end
+
+# @param file [String] filepath to a CSV containing LevelConceptDifficulty
+#   summary data.
+# @return [Hash[Integer, Array[Integer]]] mapping from level_id to
+#   LevelConceptDifficulty data.
+def read_lcd_file(file)
+  data = {}
+  CSV.foreach(file) do |row|
+    row = row.map(&:to_i)
+    data[row[1]] = row[4..13]
+  end
+  data
+end
+
+# @param best_result [Integer] the best result of the UserLevel.
+# @param attempts [Integer] the number of attempts of the UserLevel.
+# @param num_hints [Integer] the number of hints associated with the UserLevel.
+# @param num_authored_hints [Integer] the number of authored hints associated
+#   with the UserLevel.
+# @return [Float] whether the level was passed perfectly without hints.
+def perfect_without_hints(best_result, attempts, num_hints, num_authored_hints)
+  raise ArgumentError.new('FreePlay') if best_result == 30
+  return 1.0 if best_result == 100 && num_hints == 0 && num_authored_hints == 0
```

Comment:
I thought we had decided to start using `> 30` in order to accommodate diagnostic results?